File: src/main/java/algo/sort/KthSmallestPrimeFraction_786.java
package algo.sort;

import java.util.PriorityQueue;

public class KthSmallestPrimeFraction_786 {
  public static int[] kthSmallestPrimeFraction(int[] arr, int k) {
    var stop = false;
    var len = arr.length;
    var queue = new PriorityQueue<int[]>((p1, p2) -> p1[1] * p2[0] - p2[1] * p1[0]);

    for (var i = 0; i < len && !stop; i++) {
      for (var j = len - 1; j > i && !stop; j--) {
        if (queue.size() < k) {
          queue.offer(new int[]{arr[i], arr[j]});
        } else {
          var top = queue.peek();
          if (arr[i] * top[1] - arr[j] * top[0] < 0) {
            queue.poll();
            queue.offer(new int[]{arr[i], arr[j]});
          }
        }
      }
    }

    return queue.peek();
  }

  public static void main(String[] args) {
//    var fraction = kthSmallestPrimeFraction(new int[]{1, 2, 3, 5}, 3);
//    var fraction = kthSmallestPrimeFraction(new int[]{1, 7}, 1);
    var fraction = kthSmallestPrimeFraction(new int[]{1, 2, 3, 5}, 3);
    System.out.println(fraction[0] + ":" + fraction[1]);
  }
}

File: src/main/kotlin/data/structure/priorityQueue/KthSmallestPrimeFraction_786.kt
package data.structure.priorityQueue

import java.util.PriorityQueue

class KtKthSmallestPrimeFraction_786 {
    fun kthSmallestPrimeFraction(arr: IntArray, k: Int): IntArray {
        val pq = PriorityQueue<IntArray> {a, b -> a[1] * b[0] - a[0] * b[1]}  // max heap
        arr.sorted().indices.reversed().forEach { j ->
            for (i in 0..<j) {
                if (pq.size < k) {
                    pq.offer(intArrayOf(arr[i], arr[j]))
                } else {
                    val top = pq.peek()
                    if (top[0] * arr[j] < top[1] * arr[i]) {
                        break
                    }
                    // if top > i / j, then pop pop and push i / j
                    pq.poll();
                    pq.offer(intArrayOf(arr[i], arr[j]))
                }
            }
        }
        return pq.peek()
    }
}

fun main() {
    val prime = KtKthSmallestPrimeFraction_786()
    println(prime.kthSmallestPrimeFraction(intArrayOf(1,2,3,5), 3).toList())
}